### PR TITLE
drivers: i2c: Add STM32F10X slave support

### PR DESCRIPTION
**Pull request by pavlohamov**
Add i2c-slave support for STM3210X SoC series.

Signed-off-by: Pavlo Hamov <pavlo_hamov@jabil.com>

### CODE REVIEW

**Comment by ydamigos** [comment]
@pavlohamov Thanks for adding slave support to i2c_ll_stm32_v1. From a quick look the changes seems reasonable. I'll try to review it and test it the following days.

**Review by ydamigos — CHANGES_REQUESTED** [review]
@pavlohamov  How did you test it?
I tried PR on olimexino_stm32 and 96b_carbon boards using `tests/drivers/i2c/i2c_slave_api`.
It fails on both boards:
```
***** Booting Zephyr OS build zephyr-v1.14.0-2378-ge219e51a849c *****    
Running test suite test_eeprom_slave                                     
===================================================================      
starting test - test_eeprom_slave
[00:00:00.018,000] <infFAIL - test_eeprom_slave
===================================================================
Test suite test_eeprom_slave failed.
===================================================================
PROJECT EXECUTION FAILED
> main: Found I2C Master device I2C_1
[00:00:00.018,000] <inf> main: Found I2C Master device I2C_2
[00:00:00.018,000] <inf> main: Found EEPROM device EEPROM_SLAVE_0
[00:00:00.018,000] <inf> main: Found EEPROM device EEPROM_SLAVE_1
[00:00:00.018,000] <dbg> i2c_ll_stm32_v1.i2c_stm32_slave_register: i2c: slave registered
[00:00:00.018,000] <inf> main: EEPROM EEPROM_SLAVE_0 Attached !
[00:00:00.018,000] <dbg> i2c_ll_stm32_v1.i2c_stm32_slave_register: i2c: slave registered
[00:00:00.018,000] <inf> main: EEPROM EEPROM_SLAVE_1 Attached !
[00:00:00.018,000] <inf> main: Start full read. Master: I2C_1, address: 0x56
[00:00:00.018,000] <dbg> i2c_slave.eeprom_slave_write_requested: eeprom: write req
[00:00:00.018,000] <dbg> i2c_slave.eeprom_slave_write_received: eeprom: write done, val=0x0
[00:00:00.018,000] <dbg> i2c_slave.eeprom_slave_read_requested: eeprom: read req, val=0x6a
[00:00:00.018,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x69
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x68
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x67
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x66
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x65
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x64
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x63
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x62
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x61
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x39
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x38
[00:00:00.019,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x37
[00:00:00.020,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x36
[00:00:00.020,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x35
[00:00:00.020,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x34
[00:00:00.020,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x33
[00:00:00.020,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x32
[00:00:00.020,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x31
[00:00:00.020,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x30
[00:00:00.020,000] <dbg> i2c_slave.eeprom_slave_read_processed: eeprom: read done, val=0x0
[00:00:00.020,000] <inf> main: Start full read. Master: I2C_2, address: 0x54
ASSERTION FAIL [0] @ ZEPHYR_BASE/subsys/logging/log_core.c:179
        argument 0 in log message "Buffer contents are different: %s" missing log_strdup().
ASSERTION FAIL [0] @ ZEPHYR_BASE/subsys/logging/log_core.c:179
        argument 0 in log message "                           vs: %s" missing log_strdup().
***** HARD FAULT *****
  Fault escalation (see below)
***** Hardware exception *****
Current thread ID = 0x20000038 (logging)
Faulting instruction address = 0x8006592
Fatal fault in ISR! Spinning...
```

File: drivers/i2c/i2c_ll_stm32_v1.c [hunk 1]
```diff
@@ -229,6 +405,10 @@ s32_t stm32_i2c_msg_write(struct device *dev, struct i2c_msg *msg,
 	data->current.is_err = 0U;
 	data->slave_address = saddr;
 
+#if defined(CONFIG_I2C_SLAVE)
+	data->master_active = true;
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32_v1.c]: Yeah, my bad. Will add

File: drivers/i2c/i2c_ll_stm32_v1.c [hunk 2]
```diff
@@ -171,12 +207,143 @@ static inline void handle_btf(I2C_TypeDef *i2c, struct i2c_stm32_data *data)
...
+	}
+
+	stm32_i2c_disable_transfer_interrupts(dev);
+
+	LL_I2C_ClearFlag_AF(i2c);
+	LL_I2C_ClearFlag_STOP(i2c);
+	LL_I2C_ClearFlag_ADDR(i2c);
+
+	LL_I2C_Disable(i2c);
+
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32_v1.c]: Definitely should. Will add

**Comment by pavlohamov** [comment]
@ydamigos Thank you for the review. I was testing this using stm32_mini_dev (black pill) with external master device. But i even didn't thought about running the test. Seems like something complicated is hiding inside. I will fix it - will load the new patch

**Comment by pavlohamov** [comment]
@ydamigos please, take a look.
i2c_slave test is tested on stm32_min_dev_black. Also i verified that i2c master still working for 24lc16. TEST_DATA_SIZE reduced to 16 due to page write restrictions. Verified by  "run_program_read"

File: drivers/i2c/i2c_ll_stm32.c [hunk 3]
```diff
@@ -164,17 +157,15 @@ static int i2c_stm32_transfer(struct device *dev, struct i2c_msg *msg,
 		num_msgs--;
 	}
 exit:
-#if defined(CONFIG_I2C_STM32_V1)
-	LL_I2C_Disable(i2c);
```
Comment (ydamigos) [inline drivers/i2c/i2c_ll_stm32.c]: Does polling mode works after removing it?

**Review by ydamigos — CHANGES_REQUESTED** [review]
I'll try to review it during weekend.

**Comment by ydamigos** [comment]
@avisconti Could you also have a look?

File: drivers/i2c/i2c_ll_stm32.c [hunk 4]
```diff
@@ -164,17 +157,15 @@ static int i2c_stm32_transfer(struct device *dev, struct i2c_msg *msg,
...
-#endif
 	k_sem_give(&data->bus_mutex);
+	stm32_i2c_master_finish(dev);
```
Comment (ydamigos) [inline drivers/i2c/i2c_ll_stm32.c]: As master, could we use next_msg_flags pointer and disable I2C when there is no msg left to send?
For example: https://gist.github.com/ydamigos/8440e33d32da859aa566182ae3c6316b

**Review by ydamigos — CHANGES_REQUESTED** [review]
Using polling mode as master is broken because we don't enable I2C in i2c_ll_stm32.c

File: drivers/i2c/i2c_ll_stm32.c (hunk 3 above)
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32.c]: Oups, thanks. Now - yes.

File: drivers/i2c/i2c_ll_stm32_v1.c [hunk 5]
```diff
@@ -218,48 +505,39 @@ s32_t stm32_i2c_msg_write(struct device *dev, struct i2c_msg *msg,
...
+	    data->current.is_arlo) {
+		goto error;
+	}
 
-		if (data->current.is_err)
-			LOG_DBG("%s: ERR %d", __func__,
-				    data->current.is_err);
+	return 0;
+error:
+	if (data->current.is_arlo) {
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32_v1.c]: Yeah, good idea. Done

File: drivers/i2c/i2c_ll_stm32_v1.c [hunk 6]
```diff
@@ -218,48 +505,39 @@ s32_t stm32_i2c_msg_write(struct device *dev, struct i2c_msg *msg,
...
-
-	LL_I2C_EnableIT_EVT(i2c);
-	LL_I2C_EnableIT_ERR(i2c);
-	LL_I2C_AcknowledgeNextData(i2c, LL_I2C_ACK);
-	if (msg->flags & I2C_MSG_RESTART) {
-		LL_I2C_GenerateStartCondition(i2c);
-	}
-	LL_I2C_EnableIT_BUF(i2c);
+	stm32_i2c_enable_transfer_interrupts(dev);
+	LL_I2C_EnableIT_RX(i2c);
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32_v1.c]: My bad, ugly c-p. fixed

File: drivers/i2c/i2c_ll_stm32_v1.c [hunk 7]
```diff
@@ -168,48 +299,204 @@ static inline void handle_btf(I2C_TypeDef *i2c, struct i2c_stm32_data *data)
...
+#if defined(CONFIG_I2C_SLAVE)
+	if (data->slave_attached && !data->master_active) {
+		/* No need for a slave error function right now. */
+		return;
+	}
+#endif
+
 	if (LL_I2C_IsActiveFlag_AF(i2c)) {
 		LL_I2C_ClearFlag_AF(i2c);
-		LL_I2C_GenerateStopCondition(i2c);
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32_v1.c]: Thank you for noticing. i was thinking to generate stop in the "end" function. But missed it. Returned back to AF handler.

File: drivers/i2c/i2c_ll_stm32.c (hunk 4 above)
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32.c]: Good idea. Thank you for this example. I've combined "start/end" activity to use same functions for poll/interrupt modes.

**Comment by pavlohamov** [comment]
Verified slave mode. Also verified master mode (both: polling and interrupt modes).
Verified on stm32_min_dev_black and stm32f429i_disc1

File: drivers/i2c/i2c_ll_stm32_v1.c [hunk 8]
```diff
@@ -133,13 +298,24 @@ static inline void handle_rxne(I2C_TypeDef *i2c, struct i2c_stm32_data *data)
 			*data->current.buf = LL_I2C_ReceiveData8(i2c);
 			data->current.buf++;
 		}
+	} else {
+
+		if (data->current.flags & I2C_MSG_STOP) {
+			LL_I2C_GenerateStopCondition(i2c);
+		}
+		LL_I2C_TransmitData8(i2c, 0);
```
Comment (ydamigos) [inline drivers/i2c/i2c_ll_stm32_v1.c]: Is it needed?

**Review by ydamigos — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_ll_stm32_v1.c (hunk 8 above)
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32_v1.c]: oups. Was used as crutch in test. Thanks, removed.

**Review by ydamigos — APPROVED** [review]
LGTM. Tested on 96b_carbon.

File: drivers/i2c/i2c_ll_stm32_v1.c [hunk 9]
```diff
@@ -218,92 +550,74 @@ s32_t stm32_i2c_msg_write(struct device *dev, struct i2c_msg *msg,
...
+	if (LL_I2C_IsActiveFlag_AF(i2c)) {
+		LL_I2C_ClearFlag_AF(i2c);
+		data->current.is_nack = 1U;
+		goto error;
+	}
 
-	k_sem_take(&data->device_sync_sem, K_FOREVER);
+	if (LL_I2C_IsActiveFlag_ARLO(i2c)) {
+		LL_I2C_ClearFlag_ARLO(i2c);
+		data->current.is_arlo = 1U;
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32_v1.c]: To be consistent, please add dbg log message here

File: drivers/i2c/i2c_ll_stm32_v1.c [hunk 10]
```diff
@@ -476,12 +791,12 @@ s32_t stm32_i2c_msg_read(struct device *dev, struct i2c_msg *msg,
...
 				LL_I2C_GenerateStopCondition(i2c);
-				LOG_DBG("%s: NACK", __func__);
-
-				return -EIO;
+				data->current.is_nack = 1U;
+				goto end;
 			}
 		}
 
+		LL_I2C_ClearFlag_ADDR(i2c);
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32_v1.c]: Is that change related to your addition?

Comment (erwango) [reply drivers/i2c/i2c_ll_stm32_v1.c]: If not please do this as a separate commit.

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this addition.
Some changes requested for ease of review (even if most part of the review has already been done by @ydamigos). Main request is to split your changes into several, easier to review, smaller commits.
Also, this would help in debug if a regression was introduced.

File: boards/arm/stm32_min_dev/pinmux.c [hunk 11]
```diff
@@ -30,6 +30,13 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PB6, STM32F1_PINMUX_FUNC_PB6_I2C1_SCL},
 	{STM32_PIN_PB7, STM32F1_PINMUX_FUNC_PB7_I2C1_SDA},
 #endif /* CONFIG_I2C_1 */
+#ifdef CONFIG_I2C_2
+#ifdef CONFIG_UART_3
+#error "i2c2 & usart3 cannot use PB10/PB11 semitaniously"
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply boards/arm/stm32_min_dev/pinmux.c]: i've moved these changes to separate PR https://github.com/zephyrproject-rtos/zephyr/pull/17785 This is just outdated

File: drivers/i2c/i2c_ll_stm32.c [hunk 12]
```diff
@@ -40,8 +40,8 @@ int i2c_stm32_runtime_configure(struct device *dev, u32_t config)
 	clock = rcc_clocks.SYSCLK_Frequency;
 #else
 	clock_control_get_rate(device_get_binding(STM32_CLOCK_CONTROL_NAME),
-			(clock_control_subsys_t *) &cfg->pclken, &clock);
-#endif /* CONFIG_SOC_SERIES_STM32F3X) || CONFIG_SOC_SERIES_STM32F0X */
+			       (clock_control_subsys_t *) &cfg->pclken, &clock);
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32.c]: These format changes are made by uncrustify. Will revert them

File: drivers/i2c/i2c_ll_stm32_v1.c (hunk 10 above)
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32_v1.c]: Without this master read operation for 2 bytes will generate nack after first byte, thus last one wont be sent by the slave. So it's definitely required for proper work of i2c master. Just tell me, what should i do

File: drivers/i2c/i2c_ll_stm32.c [hunk 13]
```diff
@@ -218,21 +209,21 @@ static int i2c_stm32_init(struct device *dev)
 	case DT_I2C_1_BASE_ADDRESS:
 		LL_RCC_SetI2CClockSource(LL_RCC_I2C1_CLKSOURCE_SYSCLK);
 		break;
-#endif /* CONFIG_I2C_1 */
+#endif          /* CONFIG_I2C_1 */
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32.c]: Done automatically by uncrustify. In any case i've reverted them

File: drivers/i2c/i2c_ll_stm32_v1.c (hunk 9 above)
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32_v1.c]: done

**Comment by pavlohamov** [comment]
@erwango I've reverted indent change. Added log messages in check_errors function. But leaved clear adr flag movement. Please take a look

File: drivers/i2c/i2c_ll_stm32_v1.c (hunk 10 above)
Comment (erwango) [reply drivers/i2c/i2c_ll_stm32_v1.c]: Ok, then maybe add a comment in code to state this clearly.

**Review by erwango — CHANGES_REQUESTED** [review]
2 addtional comments, otherwise is fine

File: drivers/i2c/i2c_ll_stm32_v1.c [hunk 14]
```diff
@@ -168,48 +343,205 @@ static inline void handle_btf(I2C_TypeDef *i2c, struct i2c_stm32_data *data)
...
+#if defined(CONFIG_I2C_SLAVE)
+	if (data->slave_attached && !data->master_active) {
+		/* No need for a slave error function right now. */
+		return;
+	}
+#endif
+
 	if (LL_I2C_IsActiveFlag_AF(i2c)) {
 		LL_I2C_ClearFlag_AF(i2c);
 		LL_I2C_GenerateStopCondition(i2c);
```
[the remark this answers is not in the document]
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32_v1.c]: Sorry, i can't get your point. I thought code wont be executed for slave mode. Line 518 if (slave) .. return;
Could you be so kind to clarify?

File: drivers/i2c/i2c_ll_stm32_v1.c (hunk 10 above)
Comment (pavlohamov) [reply drivers/i2c/i2c_ll_stm32_v1.c]: done

File: drivers/i2c/i2c_ll_stm32_v1.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (erwango) [reply drivers/i2c/i2c_ll_stm32_v1.c]: My bad, had missed this. This is fine

**Review by erwango — APPROVED** [review]

**Comment by pavlohamov** [comment]
BTW: Is there any other way to retrigger "Shippable" check then force pushing the same again?

**Comment by erwango** [comment]
@pavlohamov 

> BTW: Is there any other way to retrigger "Shippable" check then force pushing the same again?

Hitting 'recheck' in github used to do that, but it has been non working for some time. Not sure it is fixed yet.
